pkg/substitution/substitution.go

```diff
@@ -180,38 +197,53 @@ func extractVariablesFromString(s, prefix string) ([]string, bool, string) {
 		groups := matchGroups(match, re)
 		for j, v := range []string{"var1", "var2", "var3"} {
 			val := groups[v]
-			// if using the dot notation
+			// If using the dot notation, the number of dot-separated components is restricted up to 2.
+			// Valid Examples:
+			//  - extract "aString" from <prefix>.aString
+			//  - extract "anObject" from <prefix>.anObject.key
+			// Invalid Examples:
+			//  - <prefix>.foo.bar.baz....
 			if j == 0 && strings.Contains(val, ".") {
-				switch prefix {
-				case "params":
-					// params can only have a maximum of two components in the dot notation otherwise it needs to use the bracket notation.
-					if len(strings.Split(val, ".")) > 0 {
-						errString = fmt.Sprintf(`Invalid referencing of parameters in %s !!! You can only use the dots inside single or double quotes. eg. $(params["org.foo.blah"]) or $(params['org.foo.blah']) are valid references but NOT $params.org.foo.blah.`, s)
-						return vars, true, errString
-					}
-				case "resources.(?:inputs|outputs)":
-					// resources can only have a maximum of 4 components.
-					if len(strings.Split(val, ".")) > 2 {
-						errString = fmt.Sprintf(`Invalid referencing of parameters in %s !!! resources.* can only have 4 components (eg. resources.inputs.foo.bar). Found more than 4 components.`, s)
-						return vars, true, errString
-					}
-					vars[i] = strings.SplitN(val, ".", 2)[0]
-				default:
-					// for backwards compatibality
-					vars[i] = strings.SplitN(val, ".", 2)[0]
+				if len(strings.Split(val, ".")) > 2 {
+					errString = fmt.Sprintf(`Invalid referencing of parameters in "%s"! Only two dot-separated components after the prefix "%s" are allowed.`, s, prefix)
+					return vars, true, errString
 				}
+				vars[i] = strings.SplitN(val, ".", 2)[0]
 				break
 			}
-			if groups[v] != "" {
+			if val != "" {
 				vars[i] = val
 				break
 			}
-
 		}
 	}
 	return vars, true, errString
 }
 
+func extractEntireVariablesFromString(s, prefix string) ([]string, bool) {
+	pattern := fmt.Sprintf(braceMatchingRegex, prefix, parameterSubstitution, parameterSubstitution, parameterSubstitution)
+	re := regexp.MustCompile(pattern)
```

Comment:
this looks like something that could panic; is there a version that returns an error instead of panicking?